### PR TITLE
Customize Collection Name

### PR DESCRIPTION
**Pull request by cassioeskelsen**
This change allows customizing the name of collections to more appropriate names in languages other than English. It also allows Python systems to access collections created by C# programs where it is already possible to customize the name.

### CODE REVIEW

**Review by ayende — APPROVED** [review]
One interesting question, though.
In C#, we allow you to customize that live, here you are passing a static dictionary.
I wonder if we ought to accept a lambda, instead of the dictionary?
Or would that not be the Python way to handle this?

**Comment by cassioeskelsen** [comment]
Hi, Ayende! Thanks for your comment!

Since DocumentConvention.default_collection_names is a dict property you can update it any way you want that returns a **iterable**.
You can also update the values with built-in function update:

```python
    def test_change_names_in_document_convention(self):
        dc = DocumentConventions(collection_names={SomeClass: 'SomeClass123'})
        dc.default_collection_names.update({Address: 'Address123'})
        dc.default_collection_names.update({Customer: 'Customer123'})
        assert dc.get_collection_name(SomeClass) == 'SomeClass123'
        assert dc.get_collection_name(Address) == 'Address123'
        assert dc.get_collection_name(Customer) == 'Customer123'
```
I tried to follow the same way mapper assignment works to maintain cohesion in the code.

Python allows you to create some crazier ways to update the dictionary, but I don't think it's necessary as this property should be frozen after the DocumentStore initialization. However, this is by definition, nothing prevents remove this lock and at any time the collection name can be updated.

File: pyravendb/data/document_conventions.py [hunk 1]
```diff
@@ -65,27 +69,46 @@ def _json_default(o):
...
+    def default_transform_type_tag_name(self, entity: Type):
+        if entity.__class__ in self._default_collection_names:
+            name = self._default_collection_names[entity.__class__]
+        else:
+            name = DocumentConventions.default_transform_plural(entity.__class__.__name__)
         count = sum(1 for c in name if c.isupper())
         if count <= 1:
-            return DocumentConventions.default_transform_plural(name.lower())
-        return DocumentConventions.default_transform_plural(name)
+            return name.lower()
```
Comment (IdanHaim) [inline pyravendb/data/document_conventions.py:100]: This is a mistake, default_transform_plural uses ` inflector.plural(name)`
`DocumentConventions.default_transform_plural(name.lower())` return the name from the package inflect,
here we just check if what `default_collection_names` or `default_transform_plural` returns and messed that up by call lower on the string.

**Comment by IdanHaim** [comment]
> Hi, Ayende! Thanks for your comment!
> 
> Since DocumentConvention.default_collection_names is a dict property you can update it any way you want that returns a **iterable**.
> You can also update the values with built-in function update:
> 
> ```python
>     def test_change_names_in_document_convention(self):
>         dc = DocumentConventions(collection_names={SomeClass: 'SomeClass123'})
>         dc.default_collection_names.update({Address: 'Address123'})
>         dc.default_collection_names.update({Customer: 'Customer123'})
>         assert dc.get_collection_name(SomeClass) == 'SomeClass123'
>         assert dc.get_collection_name(Address) == 'Address123'
>         assert dc.get_collection_name(Customer) == 'Customer123'
> ```
> 
> I tried to follow the same way mapper assignment works to maintain cohesion in the code.
> 
> Python allows you to create some crazier ways to update the dictionary, but I don't think it's necessary as this property should be frozen after the DocumentStore initialization. However, this is by definition, nothing prevents remove this lock and at any time the collection name can be updated.

**Comment by IdanHaim** [comment]
"I tried to follow the same way mapper assignment works to maintain cohesion in the code." - 
mappers came to solve the lack of reflection or types in python, we are using the mappers as ORM to our classes, and to be able to get our object just like in the C# code, as an initialized class and not just as a dictionary.

We need to add this change, but in my opinion:
1. change the name `default_collection_names` because the is not default (if we proceed with this approach).
2. we can add to convention a custom method to create the names of the collections to be able to get a custom method that needs to take into account when we create the collections names and in the lack of this method or with missed classes name we can just use the default one.

2 approach will give us a better generic solution we can build a method that will generate the collection names in any way we will choose for all or just specific classes.

File: pyravendb/data/document_conventions.py (hunk 1 above)
Comment (cassioeskelsen) [reply pyravendb/data/document_conventions.py:100]: Hi @IdanHaim 

inflector.plural returns the same result regardless of whether we pass the original or lower string as a parameter

```
>>> import inflect
>>> inflector = inflect.engine()
>>> inflector.plural('order')
'orders'
>>> inflector.plural('Order')
'Orders'
>>> inflector.plural('OrderItem')
'OrderItems'
>>> inflector.plural('orderitem')
'orderitems'
>>>
```

The default_transform_type_tag_name is intended to generate the object's key, and unless I get it wrong, Raven uses the following rules:

a) If it is a simple name (like Company), the key will be the plural name in lowercase: "companies/XXX"

b) If the name is composed (like Company Owner), the key will be the name with capital initials and the last name pluralized: "CompanyOwners"

What I do then is, if the class doesn't have a custom name, I first pluralize it and then adjust the result to the RavenDB standard. The snippet you referenced will only be called if the name is a single word.

You can a sample here:

```
    def test_transform_type_name(self):
        dc = DocumentConventions()
        assert "companies" == dc.default_transform_type_tag_name(Company("Test"))
        assert "CompanyOwners" == dc.default_transform_type_tag_name(CompanyOwner("Test Test"))
        assert "OrderItems" == dc.default_transform_type_tag_name(OrderItem(1, 1))
        assert "categories" == dc.default_transform_type_tag_name(Category("category"))
        assert "DemoStats" == dc.default_transform_type_tag_name(DemoStat())
```
https://gist.github.com/cassioeskelsen/80e0a31516b6b32ff7a8963a34b2d13d

**Comment by cassioeskelsen** [comment]
@IdanHaim The custom name is not just used for the class name. It is also used to generate the object key.
If we are going to open up the possibility of creating a custom method, it will have to take these two scenarios into account.

In my opinion, this adds unnecessary complexity since it's just defining a name and usually when we need to customize it it's precisely because we need to define a specific string.

Other libraries, such as Mongoengine (for MongoDB) also allow you to customize the class name by assigning a fixed string in the Class metadata. The only caveat I make in this case is that it's not interesting to put information from the infrastructure layer inside the domain classes.

**Comment by ayende** [comment]
How about when we deal with raw dictionaries?

```
session.store({ "Name": "Oren", "Valid": true })
```

And in the store, we can define it as something like:

```
store.conventions.get_docuemnt_collection = lambda entity:  
      if "Name" in entity:
            return "Users"
      return None # default to usual.
```

Or something like this

File: pyravendb/data/document_conventions.py (hunk 1 above)
Comment (IdanHaim) [reply pyravendb/data/document_conventions.py:100]: I understand you are correct about the plural, please check the behaviour for C# client custom collection names,
do we still change the string to lower if the user has a custom collection name?
if the answer is yes you can resolve this if not the if statement needs to go under the else and only for default.

Comment (cassioeskelsen) [reply pyravendb/data/document_conventions.py:100]: @IdanHaim yes, in my job we have a C# program with this:

```csharp

                    FindCollectionName = type =>
                    {
                        if (type == typeof(Operacao))
                            return "Operacoes";
                        if (type == typeof(CargaTabelas))
                            return "CargasTabelas";
                        (....)
                        return DocumentConventions.DefaultGetCollectionName(type);
                    }                
```

and on Raven database:
Collection: Operacoes,  keys: operacoes/xxxx        (only one word)
Collection: CargasTabelas, keys: CargasTabelas/xxxx   (two words)

PS: default_transform_type_tag_name don't change collection name, only key prefix.

Comment (ayende) [reply pyravendb/data/document_conventions.py:100]: The case rules for RavenDB says:

* Starts with a capital letter (only one) - lowercase. User => users/...
* Multiple capital words - as as "FooBar" => FooBars/...

**Comment by cassioeskelsen** [comment]
@ayende Today we can't store raw dictionaries with Python client 'cause they are unashable. We have same problem with dataclass, I mentioned this problem in this post: https://eskelsen.medium.com/30-dias-usando-o-ravendb-com-o-python-o-que-já-aprendi-fafe4a0dbefd (sorry, Portuguese only).

Anyway, I don't believe anyone wants to customize the name of a key in a dictionary or even in a collection through DocumentStore/DocumentConventions as they can be freely defined in the dictionary/class creation.

Python doesn't have the same concept of lambda as C#. What maybe you could do is associate it with a function, similar to what was done for mappers.

I'll think of some alternative then and submit another PR. I'm going to close this one because I don't know when I'll be able to do this.

**Comment by ayende** [comment]
About the `dict` or `@dataclass` - I wonder if we can use the `id()` function here instead. So in `_documents_by_entity`, instead of storing the entity itself as a key, we'll use the `id()` to store it.
That should make things a lot easier all around, no?

**Comment by IdanHaim** [comment]
@ayende  we already store the id as a key in `_documents_by_id`

**Comment by ayende** [comment]
Different context.
I'm talking about the `id()` function of Python, which returns the memory location of an object.
We can use that as reference equality, like we do in C#

**Comment by IdanHaim** [comment]
I read the article @cassioeskelsen sent, now I understand what you meant by Dataclass,
it will help for non-hashable objects, we need to make a few changes but it could work.

**Pull request closed without merging**